### DISCUSSION TITLE
about recover

### DISCUSSION THREAD

**Discussion by Salma-Mao** [idaholab/moose#21465] [Q&A General] [answered]
When recovering my model I get the following error:

```
*** Info ***
Using stress_copper_out_cp/7249 for recovery.

*** ERROR ***
/usr/local/workspace/stress_copper/stress_copper.i:136: (BCs/convention_surfcae/boundary):
    the following side sets (ids) do not exist on the mesh: powder_top (9)

    MOOSE distinguishes between "node sets" and "side sets" depending on whether
    you are using "Nodal" or "Integrated" BCs respectively. Node sets corresponding
    to your side sets are constructed for you by default.

    Try setting "Mesh/construct_side_list_from_node_list=true" if you see this error.
    Note: If you are running with adaptivity you should prefer using side sets.

```
I added the ''construct_side_list_from_node_list=true" in mesh. However, get the same error? what is the wrong?

  

 [Outputs]
  file_base = stress_copper_out
  execute_on = 'timestep_end'
  csv = true
  exodus = true
  type = Checkpoint
  []

**Comment by GiudGiud** [comment, marked as answer]
Did you have this side set in the initial simulation?
What does your Mesh block look like? 
You can use the mesh in the _cp folder of the checkpoint to restart btw

**Reply by Salma-Mao** [reply]
**1.0 Did you have this side set in the initial simulation? 
          What does your Mesh block look like?**
```
[Mesh]
[./total]
  type = GeneratedMeshGenerator
  dim = 3
  xmin = 0
  xmax =8.0e-4
  nx = 40

 ymin = 0
 ymax = 6.0e-4
 ny = 30

  zmin =0
  zmax = 3.3e-4
  nz = 22

[../]
[./substrate]
  type = SubdomainBoundingBoxGenerator
  input = total
  block_id = 1
  bottom_left = '0 0 0'
  top_right = '8.0e-4 6.0e-4 1.5e-4'
[../]
[./substrate_1]
  type = SubdomainBoundingBoxGenerator
  input = substrate
  block_id = 2
  bottom_left = '0 0 1.5e-4'
  top_right = '8.0e-4 6.0e-4 3.0e-4'
[../]
[./powder]
 type = SubdomainBoundingBoxGenerator
 input = substrate_1
 block_id = 3
bottom_left = '0 0 3.0e-4'
top_right = '8.0e-4 6.0e-4 3.3e-4'
[../]
[./powder_re]
 type = RefineBlockGenerator
  input = powder
  block ='2 3'
  refinement = 1
  enable_neighbor_refinement = false

[../]
[./powder_boundary_top]
      type = SideSetsAroundSubdomainGenerator
      input = powder_re
     block = 3
     normal = '0 0 1'
     new_boundary = 'powder_top'
 [../]
 []
```
**2.0 You can use the mesh in the _cp folder of the checkpoint to restart btw**

When I use the mesh in the _cp folder, I get the same error.

**Reply by GiudGiud** [reply]
The sideset generates fine for me so that isnt the problem.
The mesh restart should keep it. I just tried and it worked fine. 

This syntax is wrong btw:
```
[Outputs]
file_base = stress_copper_out
execute_on = 'timestep_end'
csv = true
exodus = true
type = Checkpoint
[]
```

maybe you mean
```
[Outputs]
  file_base = stress_copper_out
  execute_on = 'timestep_end'
  csv = true
  exodus = true
  checkpoint = true
[]
```

**Reply by GiudGiud** [reply]
what I ran to check:
```
[Mesh]
    [./total]
    type = GeneratedMeshGenerator
    dim = 3
    xmin = 0
    xmax =8.0e-4
    nx = 40
  
    ymin = 0
    ymax = 6.0e-4
    ny = 30
  
    zmin =0
    zmax = 3.3e-4
    nz = 22
  [../]
  [./substrate]
    type = SubdomainBoundingBoxGenerator
    input = total
    block_id = 1
    bottom_left = '0 0 0'
    top_right = '8.0e-4 6.0e-4 1.5e-4'
  [../]
  [./substrate_1]
    type = SubdomainBoundingBoxGenerator
    input = substrate
    block_id = 2
    bottom_left = '0 0 1.5e-4'
    top_right = '8.0e-4 6.0e-4 3.0e-4'
  [../]
  [./powder]
    type = SubdomainBoundingBoxGenerator
    input = substrate_1
    block_id = 3
    bottom_left = '0 0 3.0e-4'
    top_right = '8.0e-4 6.0e-4 3.3e-4'
  [../]
  [./powder_re]
    type = RefineBlockGenerator
    input = powder
    block ='2 3'
    refinement = 1
    enable_neighbor_refinement = false
  [../]
  [./powder_boundary_top]
    type = SideSetsAroundSubdomainGenerator
    input = powder_re
    block = 3
    normal = '0 0 1'
    new_boundary = 'powder_top'
  [../]
[]

[BCs]
  [test]
    type = DirichletBC
    boundary = powder_top
    variable = u
    value = 1
  []
[]

[Variables]
  [u]
  []
[]

[Kernels]
  [d]
    type = Diffusion
    variable = u
  []
[]

[Executioner]
  type = Steady
[]

[Outputs]
  checkpoint = true
[]

```

THEN for the checkpoint:
```
[Mesh]
  [fmg]
    type = FileMeshGenerator
    file = 'test_out_cp/0001_mesh.cpr'
  []
[]

[BCs]
  [test]
    type = DirichletBC
    boundary = powder_top
    variable = u
    value = 1
  []
[]

[Variables]
  [u]
  []
[]

[Kernels]
  [d]
    type = Diffusion
    variable = u
  []
[]

[Problem]
  restart_file_base = 'test_out_cp/LATEST'
[]

[Executioner]
  type = Steady
  nl_abs_tol = 1e-4'
  l_tol = 1e-1 # just an example!
[]

[Outputs]
  checkpoint = true
[]
```

**Reply by Salma-Mao** [reply]
**I find the problem is BCs.
I run the case.**


 [Mesh]
 [./total]
  type = GeneratedMeshGenerator
  dim = 3
  xmin = 0
xmax =8.0e-4
nx = 40
ymin = 0
ymax = 6.0e-4
ny = 30
zmin =0
zmax = 3.3e-4
nz = 22
[../]
[./substrate]
type = SubdomainBoundingBoxGenerator
input = total
block_id = 1
bottom_left = '0 0 0'
top_right = '8.0e-4 6.0e-4 1.5e-4'
[../]
[./substrate_1]
type = SubdomainBoundingBoxGenerator
input = substrate
block_id = 2
bottom_left = '0 0 1.5e-4'
top_right = '8.0e-4 6.0e-4 3.0e-4'
[../]
[./powder]
type = SubdomainBoundingBoxGenerator
input = substrate_1
block_id = 3
bottom_left = '0 0 3.0e-4'
top_right = '8.0e-4 6.0e-4 3.3e-4'
[../]
[./powder_re]
type = RefineBlockGenerator
input = powder
block ='2 3'
refinement = 1
enable_neighbor_refinement = false
[../]
[./powder_boundary_top]
type = SideSetsAroundSubdomainGenerator
input = powder_re
block = 3
normal = '0 0 1'
new_boundary = 'powder_top'
[../]
[]
[Kernels]
[./time]
type = ADHeatConductionTimeDerivative
block = '1 2 3'
variable = temp
[../]
[./heat_conduct]
type = ADHeatConduction
variable = temp
block = '1 2 3'
thermal_conductivity = thermal_conductivity
[../]
[./heat_source]
type = ADMatHeatSource
block = '2 3'
material_property = volumetric_heat
variable = temp
[../]
[]

[BCs]
[./convention_surfcae]
type = ADConvectiveHeatFluxBC
variable = temp
boundary = 'powder_top'
T_infinity = 300.0
heat_transfer_coefficient = 110
[../]
[]
[Materials]
[./heat]
type = ADHeatConductionMaterial
specific_heat = 603
thermal_conductivity = 1000
[../]
[./density]
type = ADGenericConstantMaterial
prop_names = 'density'
prop_values = '4430'
[../]
[./volumetric_heat]
type = FunctionPathEllipsoidHeatSource
rx = 1
ry = 1
rz = 1
power = 1000
efficiency = 0.5
factor = 2
function_x= path_x
function_y= path_y
function_z= path_z
[../]
[]

[Functions]
[./path_x]
type = ParsedFunction
value = 0.7*t+1.0e-4
[../]
[./path_y]
type = ParsedFunction
value = 3.0e-4
[../]
[./path_z]
type = ParsedFunction
value = 3.3e-4
[../]
[]

[Postprocessors]
[temp_max]
type = ElementExtremeValue
variable = temp
[]
[temp_min]
type = ElementExtremeValue
variable = temp
value_type = min
[]
[temp_avg]
type = ElementAverageValue
variable = temp
[]
[]

[Preconditioning]
[./full]
type = SMP
full = true
[../]
[]

[Executioner]
type = Transient
solve_type = PJFNK

nl_rel_tol = 1e-6
nl_abs_tol = 1e-6

petsc_options_iname = '-ksp_type -pc_type -pc_factor_mat_solver_package'
petsc_options_value = 'preonly lu superlu_dist'

l_max_its = 100

end_time = 1
dt = 1e-4
dtmin = 1e-5
[]

[Outputs]
execute_on = 'timestep_end'
csv = true
exodus = true
checkpoint = true
[]

**The recover**


[Problem]
restart_file_base = 'test_recover_out_cp/LATEST'
[]

[Mesh]
[./gm]
type = FileMeshGenerator
file = 'test_recover_out_cp/0017_mesh.cpr'
[../]
[]

[Variables]
[./temp]
initial_condition = 300
[]
[]

[Kernels]
[./time]
type = ADHeatConductionTimeDerivative
block = '1 2 3'
variable = temp
[../]
[./heat_conduct]
type = ADHeatConduction
variable = temp
block = '1 2 3'
thermal_conductivity = thermal_conductivity
[../]
[./heat_source]
type = ADMatHeatSource
block = '2 3'
material_property = volumetric_heat
variable = temp
[../]
[]

[BCs]
[./convention_surfcae]
type = ADConvectiveHeatFluxBC
variable = temp
boundary = 'powder_top'
T_infinity = 300.0
heat_transfer_coefficient = 110
[../]
[]
[Materials]
[./heat]
type = ADHeatConductionMaterial
specific_heat = 603
thermal_conductivity = 1000
[../]
[./density]
type = ADGenericConstantMaterial
prop_names = 'density'
prop_values = '4430
[../]
[./volumetric_heat]
type = FunctionPathEllipsoidHeatSource
rx = 1
ry = 1
rz = 1
power = 1000
efficiency = 0.5
factor = 2
function_x= path_x
function_y= path_y
function_z= path_z
[../]
[]

[Functions]
[./path_x]
type = ParsedFunction
value = 0.7*t+1.0e-4
[../]
[./path_y]
type = ParsedFunction
value = 3.0e-4
[../]
[./path_z]
type = ParsedFunction
value = 3.3e-4
[../]
[]

[Postprocessors]
[temp_max]
type = ElementExtremeValue
variable = temp
[]
[temp_min]
type = ElementExtremeValue
variable = temp
value_type = min
[]
[temp_avg]
type = ElementAverageValue
variable = temp
[]
[]

[Preconditioning]
[./full]
type = SMP
full = true
[../]
[]

[Executioner]
type = Transient
solve_type = PJFNK

nl_rel_tol = 1e-6
nl_abs_tol = 1e-6

petsc_options_iname = '-ksp_type -pc_type -pc_factor_mat_solver_package'
petsc_options_value = 'preonly lu superlu_dist'

l_max_its = 100

end_time = 1
dt = 1e-4
dtmin = 1e-5
[]

[Outputs]
execute_on = 'timestep_end'
csv = true
exodus = true
checkpoint = true
[]

**I get the same error.**
*** ERROR ***
/usr/local/workspace/test_recover/test_recover.i:41: (BCs/convention_surfcae/boundary):
the following side sets (ids) do not exist on the mesh: powder_top (7)

MOOSE distinguishes between "node sets" and "side sets" depending on whether
you are using "Nodal" or "Integrated" BCs respectively. Node sets corresponding
to your side sets are constructed for you by default.

Try setting "Mesh/construct_side_list_from_node_list=true" if you see this error.
Note: If you are running with adaptivity you should prefer using side sets.

**Reply by GiudGiud** [reply]
I just run your input file and it ran fine. 

You must be using either an old MOOSE or you have an old mesh file in the checkpoint folder.

**Reply by Salma-Mao** [reply]
I can't understand the 'You must be using either an old MOOSE or you have an old mesh file in the checkpoint folder.' 

In my case, I create **test_recover_1.i** for recovering and the **test_recover.i** is original file. The case is shown 
https://github.com/idaholab/moose/discussions/21465#discussioncomment-3061936.

![image](https://user-images.githubusercontent.com/54804946/177491308-c0a0da82-241e-4e79-b9df-0da9e948a944.png)

The checkpoint folder is shown below.

![image](https://user-images.githubusercontent.com/54804946/177491576-6a2d08d4-737d-4df5-8927-649e24579b8c.png)

When I run to vrecover using  mpirun -n 10 ./test_recover-opt -i test_recover_1.i --recover, I get the error.

*** ERROR ***
ERROR: Neither one of the following files can be located:
        '_mesh.cpr/1/header.cpr' nor
        '_mesh.cpr'
If you are running a parallel job, double check that you've created a split for 1 ranks.
Note: One of paths above may refer to a valid directory on your system, however we are attempting to read a valid header file.

**Reply by GiudGiud** [reply]
I think you ll want to pass the name of the filebase for the recover file, because the name of the input file changed between your two runs
```  --recover [file_base]                             Continue the calculation.  If file_base is omitted then the most recent recovery file will be utilized```

**Reply by Salma-Mao** [reply]
Based on the [recover](https://github.com/idaholab/moose/tree/next/test/tests/outputs/recover) , it can recover and run fine without [SideSetsAroundSubdomainGenerator](https://mooseframework.inl.gov/source/meshgenerators/SideSetsAroundSubdomainGenerator.html) . However, when [SideSetsAroundSubdomainGenerator](https://mooseframework.inl.gov/source/meshgenerators/SideSetsAroundSubdomainGenerator.html) or [SideSetsBetweenSubdomainsGenerator](https://mooseframework.inl.gov/source/meshgenerators/SideSetsBetweenSubdomainsGenerator.html) is added, I get the error.

   *** Info ***
    Using ex_12_dir_dir_cp/0006 for recovery.

   *** ERROR ***
   /usr/local/workspace/ex_12/ex_12_recover.i:116: (BCs/convention_surfcae_bottom/boundary):
    the following side sets (ids) do not exist on the mesh: powder_bottom (8)

   MOOSE distinguishes between "node sets" and "side sets" depending on whether
   you are using "Nodal" or "Integrated" BCs respectively. Node sets corresponding
   to your side sets are constructed for you by default.

   Try setting "Mesh/construct_side_list_from_node_list=true" if you see this error.
   Note: If you are running with adaptivity you should prefer using side sets.

**Reply by GiudGiud** [reply]
Did you delete the recover files in between those tries? If the code tries to recover off of a run without that generator, then the sideset is not included. 

When I ran your input files earlier this month, it worked well with SideSetsAroundSubdomainGenerator.

**Reply by Salma-Mao** [reply]
When I run my input file with SideSetsAroundSubdomainGenerator, I get a lot of error. I don't solve the problem. Could you please share with me the complete file you run runtime instructions?

**Reply by GiudGiud** [reply]
Two scripts are attached. 
run the test.txt first to create the checkpoint, then the recover script to test recovery
[test.txt](https://github.com/idaholab/moose/files/9089511/test.txt)
[recover.txt](https://github.com/idaholab/moose/files/9089512/recover.txt)
